app/models/qa/iri_template/variable_map.rb

```diff
@@ -0,0 +1,79 @@
+# Provide access to iri template variable map configuration.
+module Qa
+  module IriTemplate
+    class VariableMap
+      TYPE = "IriTemplateMapping".freeze
+      attr_reader :variable
+      attr_reader :default
+
+      # @param [Hash] map configuration hash for the variable map
+      # @option map [String] :variable name of the variable in the template (e.g. {?query} has the name 'query')
+      # @option map [String] :property always "hydra:freetextQuery" # TODO what other values are supported and what do they mean
+      # @option map [True | False] :required is this variable required
+      # @option map [String] :default value to use if a value is not provided in the request
+      def initialize(map)
+        @variable = extract_variable(config: map)
+        @required = extract_required(config: map)
+        @default = extract_default(config: map)
+        @property = map.fetch(:property, 'hydra:freetextQuery')
+      end
+
+      # Is this variable required?
+      # @returns true if required; otherwise, false
+      def required?
+        @required
+      end
+
+      # TODO: When implementing more complex query substitution, simple_value is used when template url specifies variable as {var_name}.
```

Comment:
I'm trying to understand what these two routines mean.  In the larger context does the MappingValue or the context tell how the output should be?  Aka when I construct a VariableMap does it always get evaluated the same way or does the evaluation depend change as it is called.  Things in the construction like property make me feel like it is defined when you instantiate, but these methods make me feel like it is the users choice.

Reply:
I'm going to rethink this.  It is defined by the configured mapping.  I agree that having two methods pushes that knowledge off onto the caller even though we should already know whether it is simple or parameter based on the configuration.

Reply:
Ok, here is the problem for the IriTemplate::VariableMap class.  It does not know whether a variable is defined in the Template as simple (e.g. `{var_name}`) or parameter (e.g. `{?var_name}`).  The template determines that...

```
      "template": "http://id.loc.gov/authorities/{subauth}?{?uri}",
```
* `{subauth}` is simple and returns value only
* `{?uri}` is parameter and returns parameter=value

IriTemplate::UrlConfig holds the template and the array of mappings.

IriTemplateService pulls them together to construct the values.  So the way it is written in this PR (albeit commented out for now), the service checks for patterns...

* pattern `{key}` calls simple_value on the variable map
* pattern `{?key}` calls parameter_value on the variable map

In the original implementation, I did this incorrectly.  I haven't been able to find good documentation on IRI Templates, and did some reading between the lines that was incorrect.  I've since found more information (still not great) that indicated the difference in value and parameter substitutions.  In the original implementation, I used `{?key}` as a value substitution only.

In current configurations, templates are configured incorrectly to always use `{?key}`.  So following the incorrect approach, the template example given above would be configured as...

```
      "template": "http://id.loc.gov/authorities/{?subauth}?uri={?uri}",
```
* `{?subauth}` returns value only
* `{?uri}` returns value only  (thus the template has `uri={?uri}`

I am hoping to correct this as part of the refactor process.